File: No_1054/Distant Barcodes.cpp
class Solution {
public:
    vector<int> rearrangeBarcodes(vector<int>& barcodes) {
        priority_queue<pair<int, int>> maxHeap;
        unordered_map<int, int> hmap;
        
        for (int code : barcodes)
            hmap[code]++;
        
        for (auto t : hmap)
            maxHeap.push({t.second, t.first});
        
        vector<int> ret;
        while (!maxHeap.empty()) {
            auto t1 = maxHeap.top(); maxHeap.pop();
            ret.push_back(t1.second);
            t1.first -= 1;
            if (!maxHeap.empty()) {
                auto t2 = maxHeap.top(); maxHeap.pop();
                ret.push_back(t2.second);
                t2.first -= 1;
                if (t2.first != 0)
                    maxHeap.push(t2);
            }
            if (t1.first != 0)
                maxHeap.push(t1);
        }
        
        return ret;
    }
};

File: No_1054/Distant Barcodes.py
class Solution:
    def rearrangeBarcodes(self, barcodes: List[int]) -> List[int]:
        counter = Counter(barcodes)
        
        min_heap = [(-freq, code) for code, freq in counter.items()]
        heapq.heapify(min_heap)
        
        ret = []
        while min_heap:
            freq1, code1 = heapq.heappop(min_heap)
            ret.append(code1)
            freq1 += 1
            if min_heap:
                freq2, code2 = heapq.heappop(min_heap)
                ret.append(code2)
                freq2 += 1
                if freq2:
                    heapq.heappush(min_heap, (freq2, code2))
            
            if freq1:
                heapq.heappush(min_heap, (freq1, code1))
        
        return ret